bsonutil/indexes.go

```diff
@@ -40,41 +42,52 @@ var validIndexOptions = map[string]bool{
 // All other strings that aren't one of ["2d", "geoHaystack", "2dsphere", "hashed", "text", ""]
 // will cause the index build to fail. See TOOLS-2412 for more information.
 //
-// Note, this function doesn't convert Decimal values which are equivalent to "0" (e.g. 0.00 or -0).
-//
 // This function logs the keys that are converted.
 func ConvertLegacyIndexKeys(indexKey bson.D, ns string) {
 	var converted bool
 	originalJSONString := CreateExtJSONString(indexKey)
 	for j, elem := range indexKey {
+		indexVal := 1
+		needsConversion := false
 		switch v := elem.Value.(type) {
-		case int32, int64, float64:
-			// Only convert 0 value
-			if v == 0 {
-				indexKey[j].Value = 1
-				converted = true
-			}
+		case int32:
+			indexVal = int(v)
+			needsConversion = true
+		case int64:
+			indexVal = int(v)
+			needsConversion = true
+		case float64:
+			indexVal = int(v)
+			needsConversion = true
 		case primitive.Decimal128:
-			// Note, this doesn't catch Decimal values which are equivalent to "0" (e.g. 0.00 or -0).
-			// These values are so unlikely we just ignore them
-			zeroVal, err := primitive.ParseDecimal128("0")
-			if err == nil {
-				if v == zeroVal {
-					indexKey[j].Value = 1
-					converted = true
-				}
+			if intVal, _, err := v.BigInt(); err == nil {
+				indexVal = int(intVal.Int64())
+
+				needsConversion = true
 			}
 		case string:
-			// Only convert an empty string
 			if v == "" {
 				indexKey[j].Value = 1
 				converted = true
+			} else {
+				if intVal, err := strconv.Atoi(v); err == nil {
```

Comment:
I don't think this is necessary. The only possible string values for the old indexes were `["2d", "geoHaystack", "2dsphere", "hashed", "text", ""]`. So I don't think it was possible to have a custom string.

Even if it was possible, all custom strings would needed to be converted to 1, even "-1". The ordering logic in the server for 3.2 just uses 0 if it's a string.

Reply:
Got it. I will remove this check!. Thanks!